### PR TITLE
astcenc: Update to 4.7.0

### PR DESCRIPTION
**Pull request by akien-mga**
> The 4.5.0 release is a minor release with minor image quality improvements,
> and multiple build system quality-of-life improvements.

https://github.com/ARM-software/astc-encoder/releases/tag/4.5.0

Draft as we need to review the compile flags for the "invariant" mode, which seems to be what we use (since we don't define `ASTCENC_NO_INVARIANCE`:

https://github.com/ARM-software/astc-encoder/blob/008d43f1e57944fd6c5cb663ef2c21c4884835dd/Source/cmake_core.cmake#L231-L244

See all the changelog lines talking about invariance. I'm assuming this is something we want to ensure if possible @fire? (IIRC, this means having reproducible encodings on different systems, at the cost of some performance.)

### CODE REVIEW

**Comment by fire** [comment]
The cmake appears to be for msvc enable strict for older and precise for newer. There’s also a clang cl change.

**Comment by akien-mga** [comment]
> The cmake appears to be for msvc enable strict for older and precise for newer. There’s also a clang cl change.
> 
> For floating point math.

Yeah, and also for regular Clang and GCC:
```
                $<$<AND:${is_clang},$<VERSION_GREATER_EQUAL:$<CXX_COMPILER_VERSION>,10.0.0>>:-ffp-model=precise>
                $<${is_gnu_fe}:-ffp-contract=off>)
```
(`is_gnu_fe` is for GCC and Clang IIUC)

**Comment by fire** [comment]
Since I don’t have guidance, we can try reproducing the code.

**Comment by fire** [comment]
I slept on this. 

This is trying to make the exported ASTC conversions stable across different machines.

**Comment by akien-mga** [comment]
I added the same flags as upstream, aside from the ones for ClangCL as we don't officially support it yet, and the tooling we have in `methods.py` doesn't let me detect it easily.

**Comment by Chubercik** [comment]
What's the state of this PR? There have been 3 releases since `4.5.0`, currently sitting at [`4.7.0`](https://github.com/ARM-software/astc-encoder/releases/tag/4.7.0).

File: SConstruct [hunk 1]
```diff
@@ -639,42 +641,40 @@ if selected_platform in platform_list:
...
+        "minor": 0,
+        "patch": 0,
+        "metadata1": "",
+        "metadata2": "",
+        "date": "",
     }
-    cc_version_major = int(cc_version["major"] or -1)
-    cc_version_minor = int(cc_version["minor"] or -1)
-    cc_version_metadata1 = cc_version["metadata1"] or ""
+    env.compiler_version = (int(cc_version["major"]), int(cc_version["minor"]), int(cc_version["patch"]))
```
Comment (aaronfranke) [inline SConstruct:651]: Shouldn't this refactor be in a separate PR from the astcenc library update? Perhaps as a prerequisite?

Comment (akien-mga) [reply SConstruct:651]: It is :D  #80719 
But I'm not finding the time to polish it properly.

I might update this PR and just drop the compiler-specific tweaks, to add later.

File: modules/astcenc/SCsub [hunk 2]
```diff
@@ -40,6 +40,23 @@ env_astcenc.Prepend(CPPPATH=[thirdparty_dir])
...
+
+# Compiler options to ensure invariant build.
+# Sync with upstream `Source/cmake_core.cmake` options for ASTCENC_INVARIANCE.
+if env.msvc:
+    # FIXME: MSVC compiler version detection isn't well implemented yet.
+    pass
+#   if env.compiler_version >= (19, 30, 0):
+#       env_thirdparty.AppendUnique(CCFLAGS=["/fp:precise"])
+#   else:
+#       env_thirdparty.AppendUnique(CCFLAGS=["/fp:strict"])
```
Comment (akien-mga) [inline modules/astcenc/SCsub:52]: Will look into this later, but if someone can test on MSVC and confirm that astc encoding still seems to behave ok, I think this can be merged already.

**Comment by akien-mga** [comment]
Removed all the compiler flag stuff for now, here's the old code for future reference:
```python
# Compiler options to ensure invariant build.
# Sync with upstream `Source/cmake_core.cmake` options for ASTCENC_INVARIANCE.
if env.msvc:
    pass
    if env.compiler_version >= (19, 30, 0):
       env_thirdparty.AppendUnique(CCFLAGS=["/fp:precise"])
   else:
       env_thirdparty.AppendUnique(CCFLAGS=["/fp:strict"])
else:
    from methods import using_clang

    if using_clang(env) and env.compiler_version >= (10, 0, 0):
        env_thirdparty.AppendUnique(CCFLAGS=["-ffp-model=precise"])
    env_thirdparty.AppendUnique(CCFLAGS=["-ffp-contract=off"])
```

**Review by fire — APPROVED** [review]
You only live once merge?
